### DISCUSSION TITLE
I am using onnxruntime to infer the onnx model of float16 type, the following error occurs, how should I solve it[Build]

### DISCUSSION THREAD

**Discussion by 1615070057** [microsoft/onnxruntime#17210] [General]
### Describe the issue

I am using onnxruntime to infer the onnx model of float16 type, the following error occurs, how should I solve it[Build] 
error:
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from model_fp16.onnx failed:Node (Resize__139) Op (Resize) [ShapeInferenceError] Either `sizes` or `scales` must be provided, but not both of them

### Urgency

The project is urgent, the deadline is August 23

### Target platform

linux 64

### Build script

import onnx
from onnxconverter_common import float16

model = onnx.load("model.onnx")
model_fp16 = float16.convert_float_to_float16(model)
onnx.save(model_fp16, "model_fp16.onnx", size_threshold=2048)

### Error / output

onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from model_fp16.onnx failed:Node (Resize__139) Op (Resize) [ShapeInferenceError] Either `sizes` or `scales` must be provided, but not both of them

**Comment by snnn** [comment]
It doesn't like a build issue. Please use onnx model checker to check your model first.

**Reply by 1615070057** [reply]
yes, I checked the onnx model and reported an error: onnx.onnx_cpp2py_export.checker.ValidationError: Nodes in a graph must be topologically sorted, however input 'Resize__139_input_cast_1' of node:
name: Resize__139 OpType: Resize is not output of any previous nodes.

model.onnx does not have this problem, I use onnxconverter_common to quantize the model to float16 before this error occurs.

**Reply by snnn** [reply]
So it is about how the model is generated? You should contact the authors of the tool which generated this model.

**Reply by 1615070057** [reply]
Using any tool that quantifies the onnx model, including onnxruntime (static quantization, dynamic quantization), will have this error

**Reply by snnn** [reply]
Please tell us more about how you generate the model file. I don't know which tool you used, so I don't know who can help you. So far, it doesn't seem to relate to onnxruntime. You probably report the issue to a wrong place.

**Reply by 1615070057** [reply]
The onnx.checker verification is successful, but the onnxruntime loading model reports an error: onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from model_fp16.onnx failed:This is an invalid model. Type Error: Type 'tensor(float16 )' of input parameter (const_empty_float__132) of operator (Resize) in node (Resize__139) is invalid.

**Reply by snnn** [reply]
But you said  you "checked the onnx model and reported an error: onnx.onnx_cpp2py_export.checker.ValidationError: " So I am confused. Was it successful or not?

**Reply by 1615070057** [reply]
When quantizing model.onnx to model_float16.onnx, For operator like Resize, some of the optional inputs doesn't support float16. We need exclude fp16 conversion to them. So onnx.checker verification is successful. But still inference model in onnxruntime , an error occurred: onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from model_fp16.onnx failed:This is an invalid model. Type Error: Type 'tensor(float16 )' of input parameter (const_empty_float__132 ) of operator (Resize) in node (Resize__139) is invalid.

**Reply by tianleiwu** [reply]
@1615070057,

`Type Error: Type 'tensor(float16 )' of input parameter (const_empty_float__132) of operator (Resize) in node (Resize__139) is invalid.` means that the fp16 model is invalid. 

You may try the following since [Resize is handled specially in the conversion](https://github.com/microsoft/onnxruntime/blob/1fb2e71ddc4b26bf4169bcf7e845db3951bd169d/onnxruntime/python/tools/transformers/float16.py#L12):
```
from onnxruntime.transformers.onnx_model import OnnxModel
model = OnnxModel(onnx.load_model(fp32_onnx_model_path, load_external_data=True))
model.convert_float_to_float16(keep_io_types=False)
model.save_model_to_file(fp16_onnx_model_path, use_external_data_format=False)
```

There are many parameters you can use to fine tune the conversion: https://github.com/microsoft/onnxruntime/blob/1fb2e71ddc4b26bf4169bcf7e845db3951bd169d/onnxruntime/python/tools/transformers/onnx_model.py#L723-L736

**Comment by JIANG3330** [comment]
I have the same issue.
After converting the model to an FP16 model with the following code and then inferring with ONNX Runtime, this error occurs.

```
import onnx
from onnxconverter_common import float16

model = onnx.load("test.onnx")
model_fp16 = float16.convert_float_to_float16(model)
onnx.save(model_fp16, "test_fp16.onnx")
```

If inference is done with FP32, there is no problem.
```
import onnxruntime as ort
import numpy as np

model_path = 'test.onnx'

input_data = np.random.randn(1, 3, 488, 488).astype(np.float32)

ort_session = ort.InferenceSession(model_path, providers=['CPUExecutionProvider'])
input_name = ort_session.get_inputs()[0].name
output_name = ort_session.get_outputs()[0].name
pred_onnx = ort_session.run([output_name], {input_name: input_data})[0]
```

If inference is done with an FP16 model, an error is reported.
```
import onnxruntime as ort
import numpy as np

model_path = 'test_fp16.onnx'

input_data = np.random.randn(1, 3, 488, 488).astype(np.float16)

ort_session = ort.InferenceSession(model_path, providers=['CPUExecutionProvider'])
input_name = ort_session.get_inputs()[0].name
output_name = ort_session.get_outputs()[0].name
pred_onnx = ort_session.run([output_name], {input_name: input_data})[0]
```


[onnx_model.zip](https://github.com/user-attachments/files/16558273/onnx_model.zip)

**Reply by tianleiwu** [reply]
@JIANG3330, your fp32 model exported form paddle paddle is invalid.

<img width="367" alt="image" src="https://github.com/user-attachments/assets/5d1c3fe2-2812-411d-8113-47362ec26832">

You can see the Resize op spec says `Only one of ‘scales’ and ‘sizes’ can be specified.`:
https://onnx.ai/onnx/operators/onnx__Resize.html

It seems that FP16 behavior expected. We'll check why FP32 inference did not raise the exception.
